[incrParse] Fix bug mapping a node's location back to its location in the cached syntax tree

nathawes opened the pull request:
Also fix `Edit::intersectsOrTouchesRange` check only returning true when the ranges overlapped, rather than when they overlapped or touched. This was causing an edit that extended an existing identifier, for example, to reuse the existing node, and parse the remaining characters as a separate node in the tree.

Resolves rdar://problem/45108439

Commit:
[incrParse] Fix bug mapping a node's location back to its location in the cached syntax tree

Also fix Edit::intersectsOrTouchesRange check only returning true when the
ranges overlapped, rather than when they overlapped or 'touched'.

Resolves rdar://problem/45108439

nathawes commented:
@swift-ci please smoke test

nkcsgexi reviewed (approved):
Good catch! Does the stress tester test other mutation patterns than extending a token?

ahoppen commented:
I'd like to take a closer look at this since I know that I had to switch around some `<` and `<=` in my initial implementation as well. I think I'll get to it tomorrow morning CEST.

nathawes commented:
That would be great, thanks @ahoppen!

Here's the commit where the new -> old location computation's condition was changed from < to <= if that helps: https://github.com/apple/swift/commit/b26dd1118c76db033c8cf0f896031954ea3bec9f

ahoppen commented on include/swift/Parse/SyntaxParsingCache.h:
```diff
@@ -38,7 +38,7 @@ struct SourceEdit {
   /// or are directly adjacent to it
   bool intersectsOrTouchesRange(size_t RangeStart, size_t RangeEnd) {
-    return !(End <= RangeStart || Start >= RangeEnd);
+    return End >= RangeStart && Start <= RangeEnd;
```
Good catch!

ahoppen commented on lib/Parse/SyntaxParsingCache.cpp:
```diff
@@ -85,7 +85,7 @@ llvm::Optional<Syntax> SyntaxParsingCache::lookUp(size_t NewPosition,
   for (auto I = Edits.rbegin(), E = Edits.rend(); I != E; ++I) {
     auto Edit = *I;
-    if (Edit.End <= OldPosition) {
+    if (Edit.End < OldPosition) {
```
After some more consideration this should be 
```cpp
if (Edit.End + Edit.ReplacementLength - Edit.originalLength() <= OldPosition) {
```

## Explanation

With `<=` we don't catch the following case:

Original text: `foobar`
Change `foo` to `fo` resulting in `fobar`

The edit range is 1-4 with original length 3 and replacement length 2. 
If I now try to look up `bar`, we start with `OldPosition = NewPosition = 3`. Now we need to see if that position needs to be adjusted because of the edit (which it obviously needs to be), but we get `Edit.End (4) <= OldPosition (3)` does not hold. 

The issue is that we are comparing `OldPosition` which is still in the position space of the post-edit file and `Edit.End` which is in the position space of the pre-edit file, hence we need to translate `Edit.End` to the post-position space as well.

nathawes replied:
Ah, you're right! Ok, so this should make it correct for a single edit, but if there are multiple I think we'll need to account for the original-vs-replacement length differences of any edits earlier in the file too. I'll give that a go and update shortly.

ahoppen commented on test/incrParse/extend-identifier-at-eof.swift:
```diff
@@ -0,0 +1,4 @@
+// RUN: %incr-transfer-tree --expected-incremental-syntax-tree %S/Outputs/extend-identifier-at-eof.json %s
+
+func foo() {}
+_ = x<<<|||x>>>
```
I would add comment to this file explaining that no trailing newline should be added to the end of it.

ahoppen commented on test/incrParse/simple.swift:
```diff
@@ -13,15 +13,16 @@
-_ = <<REPLACE<6|||7>>></reparse REPLACE>
-_ = <<REPLACE_BY_LONGER<6|||"Hello World">>>
+_ = <<REPLACE<6|||7>>>
+_ = </reparse REPLACE><<REPLACE_BY_LONGER<6|||"Hello World">>>
```
Putting the `</reparse REPLACE>` before the `6` seemed really odd to me and it turns out there is another off-by-one error here, where we don't report the last unexpectedly reparsed character.

Investigating further, I noticed that we need to remove the `- 1` in `swift-syntax-test.cpp:252` plus its corresponding comment. I don't know if that's related to some of your changes on `SourceLoc` or has been a bug ever since, but it fixed the issue. After that the end repairs marker also needs to go at the end of the line.

ahoppen reviewed (changes requested):
A couple comments inline that fix some more underlying bugs that have just surfaced.

ahoppen commented:
That should be taken care of by traversing the edits in reverse, which we are doing. Unless I’m completely mistaking. 

> On 10. Oct 2018, at 20:05, Nathan Hawes <notifications@github.com> wrote:
> 
> @nathawes commented on this pull request.
> 
> In lib/Parse/SyntaxParsingCache.cpp:
> 
> > @@ -85,7 +85,7 @@ llvm::Optional<Syntax> SyntaxParsingCache::lookUp(size_t NewPosition,
>    size_t OldPosition = NewPosition;
>    for (auto I = Edits.rbegin(), E = Edits.rend(); I != E; ++I) {
>      auto Edit = *I;
> -    if (Edit.End <= OldPosition) {
> +    if (Edit.End < OldPosition) {
> Ah, you're right! Ok, so this should make it correct for a single edit, but if there are multiple I think we'll need to account for the original-vs-replacement length differences of any edits earlier in the file too. I'll give that a go and update shortly.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

nathawes commented:
Ok, I might be misunderstanding how multiple edits are meant to be understood. If I we have two edits is the second edit positioned:
1) in terms of the original file, or
2) in terms of the original file after the first edit was applied?

I was thinking it was (1), where I think doing it in reverse wouldn't help, but if it's (2) you're right.

ahoppen committed:
[incrParse] Fix bug in translation of post-edit to pre-edit positions

ahoppen committed:
[incrParse] Fix issue that did not report the first reparsed character as reparsed

ahoppen committed:
[incrParse] Adjust expected reparse region for simple test case

ahoppen committed:
[incrParse] Add warning not to add anything to the identifier-at-eof test

ahoppen commented:
All changes are  in terms of the original file. But because we are now speaking in terms of the post-edit position space, we need to traverse the edits in normal order again (not reverse order) to get the correct transformation. I extracted the logic, wrote a couple test cases for it and pushed it into this PR. While I was at it, I also addressed my other review comments. 

Sorry for pushing into your branch, but it felt like this is cleaner than creating a new PR on which you would then need to rebase.

ahoppen commented:
@swift-ci Please test

nathawes commented:
No problem about pushing into the PR and thanks for adding the unit tests! The transformation looks good to me now, too. I came up with a different solution yesterday:
```
  size_t AccumulatedShift = 0;
  for (auto I = Edits.begin(), E = Edits.end(); I != E; ++I) {
    auto Edit = *I;
    size_t Shift = Edit.ReplacementLength - Edit.originalLength();
    if (Edit.End + AccumulatedShift + Shift <= NewPosition) {
      AccumulatedShift += Shift;
    } else {
      break;
    }
   }
  return /*OldPosition =*/ NewPosition - AccumulatedShift;
```
But they're equivalent, anyway (this one just applies the previous edits and current edit on the left hand side of the `<=` comparison, while yours applies the current edit on the left hand side, and un-applies the previous edits on the right hand side).

The test failure is just the expected syntax tree json needing to be updated for the 'don't add a new line' comment Alex added in one of the tests. I'll update it shortly and merge if all goes well.

Commit:
[incrParse] Fix expected syntax tree json to account for added comment

Commit:
[incrParse] Update text-util.py to handle multiple edits and reparse ranges on the same line

It wasn't accounting for the prefix length before a reparse tag previously when
keeping track of pre_column_offset and post_column_offset.

Commit:
[gardening] convert ByteBasedSourceRange to CharSourceRange rather than SourceRange

The only client of the 'toSourceRange' method immediately constructs a
CharSourceRange from it, and ByteBasedSourceRange's EndLoc isn't the start of
the last token in the range (as SourceRange expects).

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

ahoppen commented:
We need to adjust a test case in Swift Syntax to expect the extended expected reparse range.

Please test with following pull request:
https://github.com/apple/swift-syntax/pull/18

@swift-ci Please test

ahoppen commented:
Linux test failure seems unrelated.

nathawes commented:
Please test with following pull request:
apple/swift-syntax#18

@swift-ci Please smoke test Linux
